### ISSUE TITLE
List Preview and Sorting - Manual Sort via Drag and Drop Not Working

### ISSUE THREAD

**Issue by marijak** [Smile-SA/elasticsuite#3440]
After our upgrade to Magento 2.4.6, this functionality stopped working. Previously, we were using version 2.10.15. Now, we have version 2.11.7.1, and I checked the latest version, which is 2.11.9.2 but issue is persisting here too tested locally.

### Preconditions
Magento Version : 2.4.6-p7 (CE)
ElasticSuite Version : 2.11.7.1
Production

### Steps to reproduce

1. Upgrade Magento to version 2.4.6-p7
2. Navigate to the backend and try to use the drag-and-drop feature for manual sorting
3. Attempt to use the manual sorting options below the items

### Expected result

1. The drag-and-drop functionality should be available to reorder items.
2. The manual sorting option should be visible below the items for easy management.

### Actual result

1. The drag-and-drop feature for manual sorting with 3dots is missing from the backend.
2. The manual sorting options below the items are also unavailable.

![sortOrder](https://github.com/user-attachments/assets/4b6322a3-6b9f-4798-9616-b4946c7e0eab)

**Comment by Bashev** [comment]
Manual sorting works only when sorting is ... "position".

**Comment by marijak** [comment]
Hi @Bashev, yes, I see in the code of vendor/smile/elasticsuite/src/module-elasticsuite-virtual-category/Model/Preview.php, but the values I have in the backend are not equal to "position" (I don't have the option with the value "position").

I have the following values in the backend:

Here are the labels and corresponding values:

Relevance - relevance_direction_desc
Position - Ascending - position_direction_asc
Position - Descending - position_direction_desc
Product Name - Ascending - name_direction_asc
Product Name - Descending - name_direction_desc
Price - Ascending - price_direction_asc
....


Question:
So, what do you recommend to resolve this at a global level and ensure that manual sorting will work on the categories in the end? Is this the only file related to it?

We want to have the possibility of manual sorting, regardless of the setting for position.
If we have "reposition by manual" set to yes, we need to reorder only those products manually, no matter the position setting.

What code changes are needed for this? I thought it should be working by default after installing your module?

**Comment by vahonc** [comment]
Hi @marijak,

Could you specify from where came these values for `Default Product Listing Sort By` dropdown?

```
Relevance - relevance_direction_desc
Position - Ascending - position_direction_asc
Position - Descending - position_direction_desc
Product Name - Ascending - name_direction_asc
Product Name - Descending - name_direction_desc
Price - Ascending - price_direction_asc
```

I'm pretty sure that it is not Magento's default staff behavior. Maybe it's because you use a third-party module, cause from-out-the box Magento has only three options in this dropdown:

![Screenshot from 2024-11-12 17-20-51](https://github.com/user-attachments/assets/400f5c54-d3b0-4d48-b527-3c9b2fd1b839)

> I thought it should be working by default after installing your module?

Of course, it will work if you do not customize the default Magento stuff or install some 3rd party modules that can effect on this.

What was the version of Magento before the update?

BR,
Vadym

**Comment by marijak** [comment]
Hi, we are using a third-party extension for these additional options:

https://github.com/magesuite/sorting

**Comment by vahonc** [comment]
@marijak,

Compatibility with third-party modules is outside the scope of our Open Source support policy provided here. There is not much we can do for your issue. You'll have to choose between our module and the second one or add customization by yourself (by coding).

BR,
Vadym

**Comment by marijak** [comment]
Ok, I will try now with the third-party module disabled. I am able to manage the sort from the backend, but I don't see the changes on the frontend in my test environment. Do I need to reindex?

On the category list page, I don’t see any change in the sort order; they are as they were before

![image](https://github.com/user-attachments/assets/41108c13-9218-4581-879f-e229a657cb5d)

![image](https://github.com/user-attachments/assets/1d4138e8-6c40-4d8c-b121-066d50be8641)

![image](https://github.com/user-attachments/assets/668bff40-8ac2-4da0-bd92-0751a69429fd)

**Comment by vahonc** [comment]
@marijak 

After that, you select `Position` as the value for `Default Product Listing Sort By` the `Manual Sort` option will be visible below the items, and with the drag-and-drop functionality you will be able to reorder items. Don't forget to Save your category. Then of course you need to run reindex.

BR,
Vadym

**Comment by marijak** [comment]
Yes, I tried, and I see the module is installed via Composer: creativestyle/magesuite, and elasticsuite Smile is part of it
![image](https://github.com/user-attachments/assets/86deb170-41e4-40e0-80b0-5fad57698621)

Can you provide me with the steps to install only the Smile Elastic clean installation so I can test the manual sort from the beginning?

**Comment by vahonc** [comment]
@marik,

First of all, you should remove `creativestyle/magesuite` from your test environment to have ability to clean test the manual sort from the beginning:
`$ composer remove creativestyle/magesuite`

then you need to install Elasticsuite:
`$ composer require smile/elasticsuite ~2.11.0` it will install the latest version of our module.

BR,
Vadym

**Comment by marijak** [comment]
do you have a demo version so i can see how it works on clean installation?

**Comment by marijak** [comment]
I found this: https://demo.magento-elastic-suite.io/admin but could not find the credentials

**Comment by romainruaud** [comment]
Hi @marijak 

this is obvious that it's not working anymore because of this : https://github.com/Smile-SA/elasticsuite/blob/2.11.x/src/module-elasticsuite-virtual-category/Model/Preview.php#L152

If the sort order is not "position", the drag and drop is disabled.

In magesuite, the sort order can be either "position_direction_asc" or "position_direction_desc" : https://github.com/magesuite/sorting/blob/1.x/Model/Sorting/Options.php

This is, imho, not needed anymore since we allow to manage the sort direction with ElasticSuite now.

I would recommend to remove/disable the magesuite/sorting module because all it does (offering a way to control the position) is already existing in ElasticSuite now.

Regards

**Comment by marijak** [comment]
Hi, yes, thank you for your answer and support. I tried disabling it, but then other things stopped working. I cannot disable or uninstall only the magesuite/sorting module because it is part of a package: composer require creativestyle/magesuite.

There are a lot of issues in our system after disabling it, so I am unable to check further.

Anyway, if we want to control this only from Smile Elasticsearch, what changes are necessary? If we remove this code from Preview.php, we can see the control for manual sorting in the backend, but again, no changes are visible after performing actions on the frontend.

I have also opened a ticket with the Magesuite support team, but I'm still waiting for a response.
https://github.com/magesuite/magesuite/issues/126

Do you know where the entire logic for the pre-sort is located and which files are responsible? Is it complex? This will help us understand what we need to do if we want to use this feature in the future.

**Comment by romainruaud** [comment]
> Anyway, if we want to control this only from Smile Elasticsearch, what changes are necessary? If we remove this code from Preview.php, we can see the control for manual sorting in the backend, but again, no changes are visible after performing actions on the frontend.

Imho you could just create your own `di.xml` file somewhere and disable all the magesuite/sorting plugins located there : 

https://github.com/magesuite/sorting/blob/1.x/etc/di.xml

So your file should look like this : 

```
<?xml version="1.0"?>
<config xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:noNamespaceSchemaLocation="urn:magento:framework:ObjectManager/etc/config.xsd">

    <type name="Magento\Catalog\Model\Config\Source\ListSort">
        <plugin name="replace_with_options_containing_sorting_direction" disabled="true"/>
    </type>

    <type name="Magento\Catalog\Model\Category\Attribute\Source\Sortby">
        <plugin name="replace_with_options_containing_sorting_direction" disabled="true"/>
    </type>

    <type name="Magento\Catalog\Model\Category">
        <plugin name="replace_available_sort_by_options" disabled="true"/>
    </type>

    <type name="Magento\Catalog\Block\Product\ProductList\Toolbar">
        <plugin name="replace_available_orders" disabled="true">
    </type>

</config>
```

**Comment by marijak** [comment]
Hi, thanks! After adding this code to the custom module, manual sorting is activated in the backend and works fine. However, on the frontend, the product order doesn't match the backend. I even switched to the Luma theme, set the default sort to "position," and reindexed, but the issue persists.


![image](https://github.com/user-attachments/assets/2b2a9bc5-e5af-41f9-b1c5-99c062c11a1f)
![image](https://github.com/user-attachments/assets/b7ed9c5b-964a-4dbf-8785-ba0f106d10de)

**Comment by romainruaud** [comment]
In your capture it looks like it's applying "position DESC" if I trust the picture of the arrow, is it the case ?

regards

**Comment by vahonc** [comment]
@romainruaud,

Nope, if the arrow is Up it means that Ascending Direction is set, and if the arrow is Down = Descending Direction.

@marijak, are you sure that you Saved your category after manual sorting? Are there no other third-party extensions that could affect that?

BR,
Vadym

**Comment by romainruaud** [comment]
OK for the arrow :)

On top of that, maybe you can attach the content of the elasticsearch query which is generated on this page.

You can see it on the system.log file after enabling the debug mode of Elasticsuite in **Store>Configuration>ElasticSuite>Base Settings > Debug mode**

**Comment by marijak** [comment]
On my local environment, the debug log shows the following:
"Elasticsearch built-in security features are not enabled. Without authentication, your cluster could be accessible to anyone."

**Comment by marijak** [comment]
I also thought that, regardless of ascending or descending order, the products set to manual sort should always appear in this order before all others in the collection. And yes, I am sure I saved the category.

**Comment by romainruaud** [comment]
Hi @marijak 

did you manage to solve your issue ?

When I asked for the content of the system.log file I was expecting to see an Elasticsearch query inside it, when you reach the search result page.

Regards
